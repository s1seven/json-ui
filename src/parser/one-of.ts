import { isArray, isBoolean, isObject, omit } from "lodash";
import { ExtractObjects, JSONSchema7Value } from "../utils/helper-types";
import { deepMerge } from "../utils/deep-merge";
import { JSONSchema7 } from "json-schema";
import { ajv } from "./ajv";
import { ErrorObject } from "ajv";

export const getOneOfProp = (
  item: JSONSchema7Value
): JSONSchema7[] | undefined => {
  if (!isObject(item) || isArray(item) || !isArray(item.oneOf)) return void 0;
  const { oneOf } = item;
  if (oneOf.findIndex(isBoolean) !== -1)
    throw new Error("booleans not supported in anyOf");
  return oneOf as JSONSchema7[];
};

export const oneOf = (item: JSONSchema7, index: number): JSONSchema7 => {
  const oneOfVal = getOneOfProp(item);
  if (!oneOfVal) return item;
  if (oneOfVal.length <= index)
    throw new Error(
      `Index ${index} out of bounds for oneOf array of length ${oneOfVal.length}`
    );
  if (!isObject(oneOfVal[index]))
    throw new Error("oneOf item is not an object");
  return deepMerge<ExtractObjects<JSONSchema7Value>>(
    oneOfVal[index] as ExtractObjects<JSONSchema7Value>,
    omit(
      item as ExtractObjects<JSONSchema7Value>,
      "oneOf"
    ) as ExtractObjects<JSONSchema7Value>
  );
};

export const oneOfOptions = (schema: JSONSchema7) =>
  (getOneOfProp(schema) ?? []).map((val, i) => val.title ?? `Option ${i}`);

export const inferOneOfOption = (
  schema: JSONSchema7,
  value: unknown
): [number, null | ErrorObject[]] => {
  const options = oneOfOptions(schema);
  let errors = null;
  const index = options.findIndex((_, i) => {
    const a = oneOf(schema, i);
    const b = ajv.compile(a);
    const c = b(value);
    errors = b.errors;
    return c;
  });
  return [index, errors];
};
