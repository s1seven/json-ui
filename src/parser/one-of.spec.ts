import { JSONSchema7 } from "json-schema";
import { oneOf, oneOfOptions } from "./one-of";

test("it matches the snapshot", async () => {
  const schema = (await import("../../fixtures/schema-with-one-of.json"))
    .default as unknown as JSONSchema7;
  const result = oneOf(schema, 1);
  expect(result).toMatchSnapshot();
});

test("oneOf options parsed correctly", async () => {
  const schema = (await import("../../fixtures/schema-with-one-of.json"))
    .default as unknown as JSONSchema7;
  const result = oneOfOptions(schema, "", 1);
  expect(result).toMatchSnapshot();
});
