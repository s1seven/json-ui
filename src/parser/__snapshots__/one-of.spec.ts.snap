// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "properties": {
    "company": {
      "type": "string",
    },
    "employees": {
      "minimum": 1,
      "type": "number",
    },
    "school": {
      "type": "string",
    },
  },
  "required": [
    "company",
    "employees",
  ],
  "title": "Person",
  "type": "object",
}
`;

exports[`oneOf options parsed correctly 1`] = `
[
  {
    "isSelected": false,
    "title": "Option 0",
  },
  {
    "isSelected": true,
    "title": "Option 1",
  },
]
`;
